woocommerce/woocommerce-android
# Improve CurrencyEditText

Closes: #5906
Depends on Fluxc PR https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2296


### Description
This PR tries to improve our `CurrencyEditText` and its layout `WCMaterialOutlinedCurrencyEditTextView`, the list of improvements added is:
1. Ability to move the cursor and edit content anywhere, previously the changes always occurred at the last position.
2. Ability to differentiate between `0` and empty state, this behavior is optional through the property `supportsEmptyState`.
3. Support of negative values, this behavior is optional via the property `supportsNegativeValues`
4. The currency handling is done now using the `prefix`/`suffix` properties of `TextInputLayout`, which makes the cursor position more accurate, instead of pointing to the currency when currency is on the right.
5. Support of two edit modes: one with full formating, and one without it. This is useful for making the editing feels more natural for some screens.

This PR doesn't change any behavior to the screens that were using the `CurrencyEditText`, meaning both `supportsEmptyState` and `supportsNegativeValues` are set to false, and on future PRs we can update each screen with the appropriate combination.
I also was planning on updating `ProductPricingFragment` to go back to using `WCMaterialOutlinedCurrencyEditTextView`, since it supports negative values now, but I decided to do it in a separate PR since this was is already big, I'll try to work on this on a separate PR, and try to fix this [bug](https://github.com/woocommerce/woocommerce-android/issues/3496), since now we can differentiate between empty states and `0`.

This is how the field behaves now:

**Simple Payments**

https://user-images.githubusercontent.com/1657201/156342883-2432eb36-32db-4b92-8452-89124cfb606a.mov 

**Other fields (without full-formating)**

https://user-images.githubusercontent.com/1657201/156342848-04dff67c-2905-49d6-bad9-9799d0417d52.mov

And this is how the UI compares in before vs after:

| Before | After |
| ----- | ----- |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 59 47" src="https://user-images.githubusercontent.com/1657201/155560718-eb5cbb89-e708-4abf-bed1-f7764b7dc190.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 49 43" src="https://user-images.githubusercontent.com/1657201/155560748-fd8828e9-9f5c-43e7-8d4d-84e450734e6b.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 17 00 06" src="https://user-images.githubusercontent.com/1657201/155560833-0d731b0c-7405-4cb6-b348-f4c3ac7fca14.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 49 11" src="https://user-images.githubusercontent.com/1657201/155560861-4753bdf4-957a-4177-bae0-3b2133a273ea.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 59 13" src="https://user-images.githubusercontent.com/1657201/155561100-59b08667-50a9-4968-bc48-6bf65fea9302.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 50 52" src="https://user-images.githubusercontent.com/1657201/155561113-e353a45c-7963-477e-8e9d-6f9db2d7543c.png"> |
| <img width="300" alt="Screen Shot 2022-02-24 at 16 58 59" src="https://user-images.githubusercontent.com/1657201/155561166-ee71b450-a2f7-4c0f-a6dc-9fc13e7b44a2.png"> | <img width="300" alt="Screen Shot 2022-02-24 at 16 50 29" src="https://user-images.githubusercontent.com/1657201/155561179-f8c8e854-0ed2-41eb-a5d5-7cfe977d3155.png">|

### Testing instructions
The `WCMaterialOutlinedCurrencyEditTextView` is currently used in "Simple Payments" and Order Creation (shipping and fees), and it's used in refund by amount (not available in the app now), so please check for any regressions in those areas.
And please if possible, test the changes with these different options:
1. Different keyboards, I tested with Gboard and Samsung Keyboard.
2. Different formatting options from WC Core, especially different options of currency positioning (please note that to force refetch the settings, you need to fully restart the app).

A remark: for order creation, please make sure to change the code manually to update `ORDER_CREATION_M2` to `true` to be able to add `shipping` and `fees`, and please note that the `fees` are not saved when added yet.
